flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/GenericHiveMetastoreCatalogUtil.java

```diff
@@ -18,32 +18,179 @@
+
+		// Table properties
+		hiveTable.setParameters(buildFlinkProperties(properties));
+		hiveTable.getParameters().putAll(EXTERNAL_TABLE_PROPERTY);
+
+		// Hive table's StorageDescriptor
+		StorageDescriptor sd = new StorageDescriptor();
+		sd.setSerdeInfo(new SerDeInfo(null, null, new HashMap<>()));
+
+		List<FieldSchema> allColumns = createHiveColumns(table.getSchema());
+
+		// Table columns and partition keys
+		CatalogTable catalogTable = (CatalogTable) table;
+
+		if (catalogTable.isPartitioned()) {
+			int partitionKeySize = catalogTable.getPartitionKeys().size();
+			List<FieldSchema> regularColumns = allColumns.subList(0, allColumns.size() - partitionKeySize);
+			List<FieldSchema> partitionColumns = allColumns.subList(allColumns.size() - partitionKeySize, allColumns.size());
+
+			sd.setCols(regularColumns);
+			hiveTable.setPartitionKeys(partitionColumns);
+		} else {
+			sd.setCols(allColumns);
+			hiveTable.setPartitionKeys(new ArrayList<>());
+		}
+
+		hiveTable.setSd(sd);
+
+		if (table instanceof CatalogTable) {
+			// TODO: [FLINK-12240] Support view related operations in GenericHiveMetastoreCatalog
+			throw new UnsupportedOperationException();
+		}
+
+		return hiveTable;
+	}
+
+	/**
+	 * Creates a CatalogBaseTable from a Hive table.
+	 *
+	 * @param hiveTable the Hive table
+	 * @return a CatalogBaseTable
+	 */
+	public static CatalogBaseTable createCatalogTable(Table hiveTable) {
+		// Table schema
+		TableSchema tableSchema = HiveCatalogBaseUtil.createTableSchema(
+				hiveTable.getSd().getCols(), hiveTable.getPartitionKeys());
+
+		// Table properties
+		Map<String, String> properties = retrieveFlinkProperties(hiveTable.getParameters());
+
+		// Table description
+		String description = properties.remove(HiveTableConfig.TABLE_DESCRITPION);
+
+		// Partition keys
+		List<String> partitionKeys = new ArrayList<>();
+
+		if (hiveTable.getPartitionKeys() != null && hiveTable.getPartitionKeys().isEmpty()) {
+			partitionKeys = hiveTable.getPartitionKeys().stream()
+								.map(fs -> fs.getName())
+								.collect(Collectors.toList());
+		}
+
+		// TODO: remove table stats from CatalogTable
+		return new GenericCatalogTable(
+			tableSchema, new TableStats(0), partitionKeys, properties, description);
+	}
+
+	/**
+	 * Create Hive columns from Flink TableSchema.
+	 */
+	private static List<FieldSchema> createHiveColumns(TableSchema schema) {
+		String[] fieldNames = schema.getFieldNames();
+		TypeInformation[] fieldTypes = schema.getFieldTypes();
+
+		List<FieldSchema> columns = new ArrayList<>(fieldNames.length);
+
+		for (int i = 0; i < fieldNames.length; i++) {
+			columns.add(
+				new FieldSchema(fieldNames[i], HiveTypeUtil.toHiveType(fieldTypes[i]), null));
+		}
+
+		return columns;
+	}
+
+	/**
+	 * Filter out Hive-created properties, and return Flink-created properties.
+	 */
+	private static Map<String, String> retrieveFlinkProperties(Map<String, String> hiveTableParams) {
```

Comment:
Just wondering if we can keep some generic properties, even though they are not added by Flink. For example, last_modified_time, etc.

Reply:
I'd rather not to mix flink properties with hive's, otherwise it will be very hard to keep them in sync considering the hive's properties are different among versions.

I do agree we can add more generic properties, such as last modified time, for Flink tables, but I believe it needs to be on its own and not depends on hive - e.g. we define Flink's own config for last modified time and decide how to store it. That can be on a different JIRA. What do you think?

Reply:
Yeah, I think it's fine to do it with a followup JIRA.

Reply:
Sounds good.